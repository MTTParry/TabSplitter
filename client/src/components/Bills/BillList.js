import React from "react";
import { useState, useEffect } from "react";
import BillForm from "./BillForm";
import BillCard from "./BillCard";

function BillList() {
  //This needs an empty array, or the whole thing breaks
  const [bills, setBills] = useState([]);
  const [billIdToEdit, setBillIdToEdit] = useState(null);

  useEffect(() => {
    fetch("/db/bills_full")
      .then((response) => response.json())
      .then((billData) => {
        setBills(billData);
      });
  }, []);

  //  const deleteP Contacts
  const deleteBill = async (bill_id) => {
    try {
      const deleteResponse = await fetch(
        `http://localhost:5005/db/bills/${bill_id}`,
        {
          method: "DELETE",
        }
      );
      if (deleteResponse.ok) {
        setBills(bills.filter((bill) => bill.bill_id !== bill_id));
      }
    } catch (error) {
      console.error(error);
    }
  };

  //PUT stuff
  // grabs the id of the post to be editted
  const editBill = (bill_id) => {
    // console.log("Bill ID", bill_id);
    setBillIdToEdit(bill_id);
  };

  const updateBill = async (updatedBillInfo) => {
    setBills((bills) => {
      const newListBills = [];
      for (let bill of bills) {
        if (bill.bill_id === updatedBillInfo.bill_id) {
          newListBills.push(updatedBillInfo);
        } else {
          newListBills.push(bill);
        }
      }
      return newListBills;
    });
    setBillIdToEdit(null);
  };

  return (
    <div className="lists" data-testid="bill_list">
      <h2> Bill List </h2>
      <ul className="Instructions">
        This is a list of each time a person covered the bill. This includes:
        <li>the subtotal</li>
        <li>tax total</li>
        <li>the tip (percentage)</li>
        <li>when the transaction happened</li>
        <li>who paid for it</li>
        <li>have they been paid back?</li>
      </ul>

      {bills.map((bill) => {
        if (bill.bill_id === billIdToEdit) {
          return (
            <BillForm
              initialBill={bill}
              onSave={updateBill}
              key={bill.bill_id}
            />
          );
        } else {
          return (<BillCard key={bill.bill_id} bill={bill} editBill={editBill} deleteBill={deleteBill} />);
        }
      })}
    </div>
  );
}

export default BillList;
