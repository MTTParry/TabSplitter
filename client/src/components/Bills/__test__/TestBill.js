// TODO import test data
const TestPayee = {
  first_name: "Samuel",
  last_name: "Biggins",
  email: "TheBigSamowski@gmail.com",
  preferred_payment_method: "PayPal",
};

const TestBill = {
  bill_id: "id500",
  transaction_date: "2022-11-16",
  subtotal: 100,
  tax_rate: 0.05,
  tax_total: 5,
  tip_rate: 20,
  tip_total: 20,
  payee: TestPayee,
  paid_up: false,
  bill_notes: "We celebrated Jane's Birthday",
  full_total: 125,
  location: "Dogfish Alehouse",
};

// TODO fill out data
const RandomBill = () => {
  return {}
};

export default TestBill;
export { TestBill, RandomBill };
