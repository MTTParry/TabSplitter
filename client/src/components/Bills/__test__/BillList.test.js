import React from 'react'
import BillList from "../BillList"
import { render, screen, cleanup } from '@testing-library/react'
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer"
import TestBill from "./TestBill";
import { act } from "react-dom/test-utils";

afterEach(cleanup);

// Basic Test
it("renders without crashing", () => {
    render(<BillList  />)
    expect(screen.getByTestId('bill_list')).toHaveTextContent("Bill List")
})

// Snapshot test (tracks snapshots? not as useful for )
it("matches snapshot", () => {
    const tree = renderer.create(<BillList />).toJSON();
    expect(tree).toMatchSnapshot()
})

// trying to mock fetch - cant figure out rn
beforeEach(() => {
  jest.spyOn(global, "fetch").mockResolvedValue({
    json: jest.fn().mockResolvedValue([TestBill]),
  });
});

afterEach(() => {
  jest.restoreAllMocks();
});

it("renders one bill correctly", async () => {
    // fetch.mockResponseOnce(() => Promise.reject("API is Down"))
    act( () => {
        render(<BillList/>)
    })
    const element = await screen.findByTestId("bill_list")
    expect(element).toBeInTheDocument();
})