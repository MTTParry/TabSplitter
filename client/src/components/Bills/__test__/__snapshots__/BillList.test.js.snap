// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="lists"
  data-testid="bill_list"
>
  <h2>
     Bill List 
  </h2>
  <ul
    className="Instructions"
  >
    This is a list of each time a person covered the bill. This includes:
    <li>
      the subtotal
    </li>
    <li>
      tax total
    </li>
    <li>
      the tip (percentage)
    </li>
    <li>
      when the transaction happened
    </li>
    <li>
      who paid for it
    </li>
    <li>
      have they been paid back?
    </li>
  </ul>
</div>
`;
